### PR TITLE
Fix crash after slept and waked app

### PR DESCRIPTION
**Pull request by AnastasiaKubova**
* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Fixed problem with application crash after switched startup mode.

1. Set startup mode as OneCollector
2. Kill the Sasquatch app
3. Relaunch it
4. Click Home to background the app
5. Wait 3 seconds
6. Wake up the app to foreground it
7. Crash

## Related PRs or issues

[AB#57535](https://msmobilecenter.visualstudio.com/web/wi.aspx?pcguid=27bf89de-9697-418a-accf-1aa125b22914&id=57535)

### CODE REVIEW

File: AppCenterDistribute/AppCenterDistribute/MSDistribute.m [hunk 1]
```diff
@@ -1108,7 +1108,7 @@ - (BOOL)openURL:(NSURL *)url {
 }
 
 - (void)applicationWillEnterForeground {
-  if ([self isEnabled] && ![MS_USER_DEFAULTS objectForKey:kMSUpdateTokenRequestIdKey]) {
+  if ([self started] && [self isEnabled] && ![MS_USER_DEFAULTS objectForKey:kMSUpdateTokenRequestIdKey]) {
```
Comment (clpolet) [inline AppCenterDistribute/AppCenterDistribute/MSDistribute.m]: There is also `canBeUsed` it' s probably covering all grounds at once. But that's the way to go.

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 2]
```diff
@@ -1785,30 +1785,44 @@ - (void)testStartUpdateWhenApplicationEnterForeground {
...
+  XCTAssertEqual(startUpdateCounter, 1);
 
   // When
   [distributeMock setEnabled:NO];
   [notificationCenterMock postNotificationName:UIApplicationWillEnterForegroundNotification object:nil];
 
   // Then
   OCMVerify([distributeMock isEnabled]);
-  XCTAssertEqual(startUpdateCounter, 0);
+  XCTAssertEqual(startUpdateCounter, 1);
```
Comment (Jamminroot) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: nit: Can use OCMVerify over lambda counter and assert checks:
```Objective-C
 OCMStub([distributeMock startUpdate]).andDo(^(__unused NSInvocation *invocation) {
    startUpdateCounter++;
  });
```

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 3]
```diff
@@ -2335,4 +2349,43 @@ - (id)mockMSPackageHash {
   return utilityMock;
 }
 
+- (void)testStartUpdateWhenEnabledButDoesNotStarted {
```
Comment (Jamminroot) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]:
```suggestion
- (void)testStartUpdateWhenEnabledButDidNotStart {
```

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 4]
```diff
@@ -2335,4 +2349,43 @@ - (id)mockMSPackageHash {
...
+  NSString *isEnabledKye = @"MSAppCenterIsEnabled";
+  [MS_USER_DEFAULTS setObject:@(YES) forKey:isEnabledKye];
+    
+  // If
+  id notificationCenterMock = OCMPartialMock([NSNotificationCenter new]);
+  OCMStub([notificationCenterMock defaultCenter]).andReturn(notificationCenterMock);
+  id distributeMock = OCMPartialMock([MSDistribute new]);
+  __block int startUpdateCounter = 0;
+  OCMStub([distributeMock startUpdate]).andDo(^(__unused NSInvocation *invocation) {
+      startUpdateCounter++;
```
Comment (Jamminroot) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: Use OCMReject over lambda and counter check

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 5]
```diff
@@ -1784,31 +1784,42 @@ - (void)testStartUpdateWhenApplicationEnterForeground {
...
-  XCTAssertEqual(startUpdateCounter, 1);
+  OCMVerify([distributeMock startUpdate]);
 
   // When
   [notificationCenterMock postNotificationName:UIApplicationWillEnterForegroundNotification object:nil];
 
   // Then
   OCMVerify([distributeMock isEnabled]);
-  XCTAssertEqual(startUpdateCounter, 2);
+  OCMVerify([distributeMock startUpdate]);
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: You are missing the count of calls here. If it is called only once, the test will pass

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 6]
```diff
@@ -2335,4 +2346,40 @@ - (id)mockMSPackageHash {
...
+
+  // Then
+  OCMReject([distributeMock startUpdate]);
+
+  // When
+  [notificationCenterMock postNotificationName:UIApplicationWillEnterForegroundNotification object:nil];
+
+  // Then
+  OCMVerify([distributeMock isEnabled]);
+  OCMReject([distributeMock startUpdate]);
```
Comment (MatkovIvan) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]: `OCMReject` has a different logic. It should be called once at the start

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 7]
```diff
@@ -2335,4 +2349,36 @@ - (id)mockMSPackageHash {
   return utilityMock;
 }
 
+- (void)testStartUpdateWhenEnabledButDidNotStart {
+  NSString *isEnabledKye = @"MSAppCenterIsEnabled";
```
Comment (jwallra) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]:
```suggestion
  NSString *isEnabledKey = @"MSAppCenterIsEnabled";
```

File: AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m [hunk 8]
```diff
@@ -2335,4 +2349,36 @@ - (id)mockMSPackageHash {
...
+  [MS_USER_DEFAULTS setObject:@(YES) forKey:isEnabledKye];
```
Comment (jwallra) [inline AppCenterDistribute/AppCenterDistributeTests/MSDistributeTests.m]:
```suggestion
  [MS_USER_DEFAULTS setObject:@(YES) forKey:isEnabledKey];
```

**Pull request merged**